### PR TITLE
[#276] Employee can add Issues

### PR DESCRIPTION
**Pull request by sandeepc19**

### CODE REVIEW

File: app/models/employee.rb [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
+  def create_corresponding_user
+    user = User.find_or_create_by(email: email)
+    self.user_id = user.id
+    save
+  end
+
+  def delete_user_and_deallocate_items
+    if active_changed? && active == false
+      items.clear()
+      user.destroy if user.present?
```
Comment (nbanerjee) [inline app/models/employee.rb]: user should be deactivated not deleted.

File: app/views/layouts/_navigation.html.erb [hunk 2]
```diff
@@ -11,38 +11,47 @@
...
-              </li>
-              <li>
-                <%= link_to 'Users', users_path %>
-              </li>
-              <li>
-                <%= link_to 'Vendors', vendors_path %>
-              </li>
-            </ul>
-          </li>
+	   <% unless current_user.admin %>
```
Comment (nbanerjee) [inline app/views/layouts/_navigation.html.erb]: use `if`

File: config/routes.rb [hunk 3]
```diff
@@ -48,4 +48,15 @@
...
   delete 'logout', to: 'sessions#destroy'
+
+  namespace :employee do
+    resources :issues, only: [:edit, :update, :new, :create, :index] do
+      member do
+        put 'set_priority'
+      end
+    end
+
+    resources :checkouts, except: [:destroy, :checkin] do
```
Comment (nbanerjee) [inline config/routes.rb]: Fix indentation

File: app/controllers/employee/checkouts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    @checkout = Checkout.new
+  end
+
+  def create
+    @checkout = Checkout.new(checkout_params)
+
+    if request.xhr?
+      @checkout.save
+    else
+      if @checkout.save
```
Comment (nbanerjee) [inline app/controllers/employee/checkouts_controller.rb]: remove nested condition.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -50,6 +51,10 @@ def get_user
   end
 
   def user_params
-    params.require(:user).permit(:email)
+    params.require(:user).permit(:email, :admin)
+  end
+
+  def authenticate_admin
```
Comment (nbanerjee) [inline app/controllers/users_controller.rb]: Move it to `ApplicationController`

File: app/controllers/employee/checkouts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
+class Employee::CheckoutsController < ApplicationController
+  before_action :get_checkout, only: [:edit, :update, :show, :checkin]
```
Comment (nbanerjee) [inline app/controllers/employee/checkouts_controller.rb]: Please don't load object in before action.

File: db/migrate/20170310114118_add_user_id_to_existing_employees.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class AddUserIdToExistingEmployees < ActiveRecord::Migration
+  def up
+    employees = Employee.where(user_id: nil, active: true)
```
Comment (nbanerjee) [inline db/migrate/20170310114118_add_user_id_to_existing_employees.rb]: Add line break

File: app/controllers/employee/issues_controller.rb [hunk 8]
```diff
@@ -0,0 +1,66 @@
+class Employee::IssuesController < ApplicationController
+  before_action :get_issue, only: [:edit, :update, :show, :set_priority]
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: Please don't load object in before action.

File: app/controllers/employee/checkouts_controller.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    elsif @checkout.update_attributes(checkout_params)
+      redirect_to :back, flash: { success: t('update') }
+    else
+      render 'edit'
+    end
+  end
+
+  def index
+    @checkouts = Checkout.includes(item: [:brand, :category]).paginate(page: params[:page]).order_desending
+    @checkouts = @checkouts.where(item_id: params[:item_id]) if params[:item_id].present?
```
Comment (nbanerjee) [inline app/controllers/employee/checkouts_controller.rb]: What is this? One employee can see only his checkouts.

File: app/controllers/employee/checkouts_controller.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
+class Employee::CheckoutsController < ApplicationController
+
```
Comment (nbanerjee) [inline app/controllers/employee/checkouts_controller.rb]: Remove this line

File: app/views/checkouts/_form.html.erb [hunk 11]
```diff
@@ -1,6 +1,12 @@
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :item_id , t('.title.item'), class: 'required' %></div>
-  <div class="col-sm-8"><%= f.select :item_id, options_for_select(Item.not_erased.active.available.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }), { prompt: t('.placeholder.select_item') }, class: 'form-control' %></div>
+  <div class="col-sm-8">
+    <% if current_user.admin %>
+      <%= f.select :item_id, options_for_select(Item.not_erased.active.available.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }), { prompt: t('.placeholder.select_item') }, class: 'form-control' %>
```
Comment (suman-kumari) [inline app/views/checkouts/_form.html.erb]: use helpers

File: app/views/issues/_form.html.erb [hunk 12]
```diff
@@ -5,7 +5,11 @@
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :items, t('.title.item') %></div>
   <div class="col-sm-8">
-    <%= f.select :item_id, options_for_select(Item.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: @issue.item_id), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
+    <% if current_user.admin %>
+      <%= f.select :item_id, options_for_select(Item.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: @issue.item_id), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
```
Comment (suman-kumari) [inline app/views/issues/_form.html.erb]: use helpers

File: app/views/issues/_form.html.erb [hunk 13]
```diff
@@ -5,7 +5,11 @@
...
+    <% else %>
+      <%= f.select :item_id, options_for_select(current_user.employee.items.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: params[:item_id]), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
```
Comment (suman-kumari) [inline app/views/issues/_form.html.erb]: use helpers

File: db/migrate/20170222061836_add_admin_to_users.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+class AddAdminToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :admin, :boolean, default: false
```
Comment (suman-kumari) [inline db/migrate/20170222061836_add_admin_to_users.rb]: add null: false for all the columns which cannot be null

File: app/controllers/employee/checkouts_controller.rb [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  def show
+    @checkout = Checkout.find(params[:id])
+  end
+
+  def create
+    @checkout = Checkout.new(checkout_params)
+
+    if request.xhr?
+      @checkout.save
+    elsif @checkout.save
```
Comment (suman-kumari) [inline app/controllers/employee/checkouts_controller.rb]: if @checkout.save
redirect_to :back, flash: { success: t('create') } unless request.xhr?
else
render :new:
end

use this code

File: app/controllers/employee/checkouts_controller.rb [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+      render 'new'
+    end
+  end
+
+  def update
+    @checkout = Checkout.find(params[:id])
+
+    if request.xhr?
+      @checkout.update_attributes(checkout_params)
+    elsif @checkout.update_attributes(checkout_params)
```
Comment (suman-kumari) [inline app/controllers/employee/checkouts_controller.rb]: use single if else block as mentioned above

File: app/controllers/employee/issues_controller.rb [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  def show
+    @issue = Issue.find(params[:id])
+  end
+
+  def create
+    @issue = Issue.new(issue_params)
+
+    if request.xhr?
+      @issue.save
+    elsif @issue.save
```
Comment (suman-kumari) [inline app/controllers/employee/issues_controller.rb]: use single if else block as mentioned above

File: app/controllers/employee/issues_controller.rb [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+      render 'new'
+    end
+  end
+
+  def update
+    @issue = Issue.find(params[:id])
+
+    if request.xhr?
+      @issue.update_attributes(issue_params)
+    elsif @issue.update_attributes(issue_params)
```
Comment (suman-kumari) [inline app/controllers/employee/issues_controller.rb]: use single if else block as mentioned above

**Comment by ssinghi** [comment]
This is poorly written code.
* Why has `employees/checkout` controller been added?
* Did anyone consider the security ramifications of 
```@issue = Issue.find(params[:id])
+
+    if @issue.update_attributes(issue_params)

**Pull request closed without merging**